server/src/helpers.ts

```diff
@@ -0,0 +1,25 @@
+/*****************************************
+NOTE
+- here contains backend helpers that are not shared with client and not supposed to run on client
+- migrate utils.js stuff here
+- unit test at test/helpers.unit.spec.js
+*****************************************/
+
+/*
+joins two parts of url into one
+part 1 must be https://domain, with optional "path" or "path/", must not include search value e.g. ?k=v
+part 2 must be a path of "path" or "/path"
+
+the function is simple and doesn't support rest parameter.
+when you have 3 pieces to join, do joinUrl( joinUrl(p1,p2), p3)
+
+cannot do below:
+- path.join() normalizes double-slash '://' to ':/' single-slash
+- url.resolve() replaces the path string after the domain, instead of appending to it
+*/
+export function joinUrl(p1: string, p2: string): string | null {
+	// p1 and p2 both shouldn't be blank string. if so, return null to alert
+	if (!p1 || !p2) return null
+	if (p1.indexOf('?') != -1) return null // search string not allowed in p1. if usecase arises can support it
```

Comment:
maybe `if(p1.indexOf('?')!=-1) return null // cannot handle search string` should be an error and throw?

or it can split using `?` and join `p2` to the first part and then re-join the second part with `?`

Reply:
gdc host names won't have such. we can do it later

Reply:
Ok, but then it should just throw in line 22, since a fetch request that uses `joinUrl()` will not work anyway when the url is `null`.

Reply:
changed to throw and covered in test

but the test cannot catch type err, e.g. when joinUrl() argument is number

Reply:
I added a test to detect non-string argument, will approve this and merge once all CI pass

Reply:
thanks! i had thought that since the function arguments are already defined as string type, it will detect and reject non-string values at runtime?

Reply:
not at runtime, `tsc` will only run static type checks at compile time, so we have to either manually write tests to validate runtime arguments/request payloads or use `typia` which can auto-generate runtime type checker function based on typescript code - this is being fixed with augen refactor